### PR TITLE
use renameat2 Linux extension syscall

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
use renameat2 Linux extension syscall

## Motivation and Context
this is a faster and safer alternative
on newer kernel versions.

## How to test this PR?
This is a test code for some performance
improvements, pending changes are related
to if possible make unversioned renames()
use this code path instead.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/os-rename_linux.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+package cmd
+
+import (
+	"golang.org/x/sys/unix"
+)
+
+// Rename2 captures time taken to call os.Rename
+func Rename2(src, dst string) (err error) {
+	defer updateOSMetrics(osMetricRename2, src, dst)(err)
+	exchange := uint(2)
```
Comment (vadmeste) [inline cmd/os-rename_linux.go]: I guess you can declare this as constant (I am not sure if golang optimizes this or not)

File: cmd/xl-storage.go [hunk 2]
```diff
@@ -2481,6 +2481,13 @@ func (s *xlStorage) RenameData(ctx context.Context, srcVolume, srcPath string, f
 		}
 		diskHealthCheckOK(ctx, err)
 
+		if !fi.Versioned && !fi.Healing() {
```
Comment (vadmeste) [inline cmd/xl-storage.go:2488]: Maybe we can calculate the current kernel version and adds it as additional check here

**Review by klauspost — APPROVED** [review]
LGTM

**Review by krisis — APPROVED** [review]

**Review by vadmeste — COMMENTED** [review]
Few comments, LGTM otherwise

File: cmd/xl-storage.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/xl-storage.go:2488]: There is no reason this has been implemented for so long, and also we have a fallback.

**Comment by harshavardhana** [comment]
So not sure why this is happening - minio-js test is failing in CI but not locally

 https://github.com/minio/minio/actions/runs/5988690422/job/16244113186?pr=17757

```
(10/14) Running minio-js tests ... FAILED in 1 minutes and 14 seconds
Warning: Cannot find any files matching pattern "test/**/*.js"
```

I have looked at the test results, there are no errors. 

```
~ npm run functional

  xml-parser
    #listObjects()
      value type casting
        ✔ should parse VersionId as string even if number is provided
        ✔ should parse Size as number


  407 passing (36s)
  3 pending
```

**Comment by harshavardhana** [comment]
@prakashsvmx can you help me investigate this?

**Comment by harshavardhana** [comment]
Found the actual error

```
{"name":"minio-js","function":"fPutObject(bucketName, objectName, filePath, metaData)","args":"bucketName:minio-js-test-c908f5e7-8d85-45c5-983b-84e4459160c3, objectName:datafile-65-MB, filePath:/tmp/datafile-65-MB","duration":171,"status":"FAIL","error":"S3Error: One or more of the specified parts could not be found. The part may not have been uploaded, or the specified entity tag may not match the part's entity tag. at Object.parseError (minio-js/src/xml-parsers.js:37:11) at parser (minio-js/src/transformers.js:121:22) at Transform._flush (minio-js/src/transformers.js:51:12) at Transform.prefinish (minio-js/node_modules/readable-stream/lib/_stream_transform.js:123:10) at Transform.emit (node:events:512:28) at prefinish (minio-js/node_modules/readable-stream/lib/_stream_writable.js:569:14) at finishMaybe (minio-js/node_modules/readable-stream/lib/_stream_writable.js:576:5) at endWritable (minio-js/node_modules/readable-stream/lib/_stream_writable.js:594:3) at Transform.Writable.end (minio-js/node_modules/readable-stream/lib/_stream_writable.js:535:22) at IncomingMessage.onend (node:internal/streams/readable:724:10) at Object.onceWrapper (node:events:626:28) at IncomingMessage.emit (node:events:524:35) at endReadableNT (node:internal/streams/readable:1378:12) at processTicksAndRejections (node:internal/process/task_queues:82:21)"}
```

**Comment by prakashsvmx** [comment]
@harshavardhana ,
 i tested , only `fPutObject` seems to fail. i could replicate the failure with 5MiB PartSize for a 65 MiB File.
The `putObject` works fine for the same 5MiB partSize and 65MiB File. 

i will investigate the reason for the same. 

Thank you :+1:

**Comment by prakashsvmx** [comment]
With the latest master of minio-js, it works fine.  
possibly fixed as part of https://github.com/minio/minio-js/pull/1160/files

**Pull request merged**